File: real_news_collector.js
const axios = require('axios');
const xml2js = require('xml2js');
const fs = require('fs');
const path = require('path');

// RSS源列表（AI科技和财经）
const RSS_SOURCES = [
  {
    name: 'TechCrunch AI',
    url: 'https://techcrunch.com/tag/ai/feed/',
    category: 'AI'
  },
  {
    name: 'MIT Technology Review AI',
    url: 'https://www.technologyreview.com/topic/artificial-intelligence/feed/',
    category: 'AI'
  },
  {
    name: 'Reuters Technology',
    url: 'https://www.reuters.com/technology/',
    category: 'Tech'
  },
  {
    name: 'Bloomberg Technology',
    url: 'https://www.bloomberg.com/technology',
    category: 'Finance'
  }
];

async function fetchRSSFeed(rssUrl) {
  try {
    console.log(`📡 获取RSS: ${rssUrl}`);
    const response = await axios.get(rssUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      },
      timeout: 10000
    });
    
    const parser = new xml2js.Parser();
    const result = await parser.parseStringPromise(response.data);
    
    const items = result.rss?.channel?.[0]?.item || [];
    return items.slice(0, 5); // 只取最新5条
  } catch (error) {
    console.error(`❌ 获取RSS失败 ${rssUrl}:`, error.message);
    return [];
  }
}

async function collectRealNews() {
  console.log('🎯 开始收集真实新闻...');
  console.log('='.repeat(50));
  
  const allNews = [];
  
  for (const source of RSS_SOURCES) {
    const items = await fetchRSSFeed(source.url);
    
    for (const item of items) {
      const news = {
        title: item.title?.[0] || '无标题',
        description: item.description?.[0] || '无描述',
        link: item.link?.[0] || '#',
        pubDate: item.pubDate?.[0] || new Date().toISOString(),
        source: source.name,
        category: source.category
      };
      
      // 过滤包含AI、科技、金融关键词的新闻
      const keywords = ['AI', 'artificial intelligence', 'machine learning', 'tech', 'finance', 'investment', 'startup'];
      const titleLower = news.title.toLowerCase();
      
      if (keywords.some(keyword => titleLower.includes(keyword.toLowerCase()))) {
        allNews.push(news);
        console.log(`📰 发现新闻: ${news.title}`);
        console.log(`   🔗 来源: ${news.source}`);
        console.log(`   📅 时间: ${news.pubDate.substring(0, 25)}`);
      }
    }
  }
  
  console.log('='.repeat(50));
  console.log(`✅ 共收集到 ${allNews.length} 条相关新闻`);
  
  return allNews;
}

function createHTMLReport(newsItems) {
  const now = new Date();
  const dateStr = now.toLocaleString('zh-CN', { 
    timeZone: 'Asia/Shanghai',
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit'
  });
  
  let html = `<!DOCTYPE html>
<html lang="zh-CN">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>真实AI科技财经新闻 - ${dateStr}</title>
    <style>
        body {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
            line-height: 1.6;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            padding: 30px;
            min-height: 100vh;
        }
        .container {
            max-width: 1000px;
            margin: 0 auto;
            background: white;
            border-radius: 20px;
            padding: 40px;
            box-shadow: 0 20px 60px rgba(0,0,0,0.3);
        }
        h1 {
            color: #2c3e50;
            border-bottom: 3px solid #3498db;
            padding-bottom: 15px;
            margin-bottom: 30px;
        }
        .news-item {
            background: #f8f9fa;
            border-radius: 15px;
            padding: 25px;
            margin-bottom: 25px;
            border-left: 5px solid #3498db;
        }
        .news-item.ai {
            border-left-color: #2ecc71;
        }
        .news-item.finance {
            border-left-color: #f39c12;
        }
        .news-title {
            font-size: 22px;
            font-weight: 600;
            color: #2c3e50;
            margin-bottom: 10px;
        }
        .news-description {
            color: #555;
            margin-bottom: 15px;
        }
        .news-meta {
            color: #7f8c8d;
            font-size: 14px;
            display: flex;
            justify-content: space-between;
        }
        .footer {
            text-align: center;
            margin-top: 40px;
            color: #7f8c8d;
            border-top: 1px solid #eee;
            padding-top: 20px;
        }
    </style>
</head>
<body>
    <div class="container">
        <h1>🤖 真实AI科技财经新闻简报</h1>
        <div style="color: #7f8c8d; margin-bottom: 30px;">
            📅 ${dateStr} · 👑 超级大大王专属 · 📰 真实新闻源收集
        </div>
`;
  
  if (newsItems.length === 0) {
    html += `
        <div style="text-align: center; padding: 50px;">
            <h2>📡 新闻收集中...</h2>
            <p>正在从TechCrunch、Reuters、Bloomberg等源收集最新AI和财经新闻</p>
            <p>请稍后查看完整新闻</p>
        </div>
    `;
  } else {
    newsItems.forEach((news, index) => {
      const categoryClass = news.category === 'AI' ? 'ai' : (news.category === 'Finance' ? 'finance' : '');
      html += `
        <div class="news-item ${categoryClass}">
            <div class="news-title">${index + 1}. ${news.title}</div>
            <div class="news-description">${news.description.substring(0, 150)}...</div>
            <div class="news-meta">
                <span>📰 来源: ${news.source}</span>
                <span>⏰ ${news.pubDate.substring(0, 16)}</span>
            </div>
        </div>
      `;
    });
  }
  
  html += `
        <div class="footer">
            <p>📡 新闻源: TechCrunch AI · MIT Tech Review · Reuters · Bloomberg</p>
            <p>🤖 生成者: NIKO AI助手 · 👑 专为超级大大王定制</p>
            <p>⚠️ 注: 这是真实新闻收集测试，内容来自公开RSS源</p>
        </div>
    </div>
</body>
</html>`;
  
  return html;
}

async function main() {
  console.log('🚀 启动真实新闻收集系统...');
  console.log('👑 用户: 超级大大王');
  console.log('🎯 兴趣: AI科技(60%) + 财经(30%) + 其他科技(10%)');
  
  try {
    // 1. 收集真实新闻
    const realNews = await collectRealNews();
    
    // 2. 生成HTML报告
    const htmlContent = createHTMLReport(realNews);
    const htmlPath = path.join(__dirname, 'real_news_report.html');
    fs.writeFileSync(htmlPath, htmlContent);
    console.log(`✅ HTML报告生成: ${htmlPath}`);
    
    // 3. 转换为图片
    console.log('📸 正在转换为图片...');
    
    // 保存新闻数据
    const data = {
      timestamp: new Date().toISOString(),
      newsCount: realNews.length,
      sources: RSS_SOURCES.map(s => s.name),
      news: realNews
    };
    
    const dataPath = path.join(__dirname, 'real_news_data.json');
    fs.writeFileSync(dataPath, JSON.stringify(data, null, 2));
    console.log(`📊 新闻数据保存: ${dataPath}`);
    
    console.log('='.repeat(50));
    console.log('🎉 真实新闻收集完成！');
    console.log(`📰 新闻数量: ${realNews.length}条`);
    console.log(`📁 输出文件:`);
    console.log(`   • ${htmlPath} (HTML报告)`);
    console.log(`   • ${dataPath} (JSON数据)`);
    
    return { htmlPath, dataPath, newsCount: realNews.length };
    
  } catch (error) {
    console.error('❌ 新闻收集失败:', error.message);
    
    // 创建错误报告
    const errorHtml = `<!DOCTYPE html>
<html>
<head><title>新闻收集错误</title></head>
<body style="padding: 40px; font-family: sans-serif;">
    <h1>⚠️ 新闻收集遇到问题</h1>
    <p>错误: ${error.message}</p>
    <p>时间: ${new Date().toLocaleString('zh-CN')}</p>
    <p>👑 用户: 超级大大王</p>
    <p>🤖 NIKO正在修复此问题...</p>
</body>
</html>`;
    
    const errorPath = path.join(__dirname, 'news_error.html');
    fs.writeFileSync(errorPath, errorHtml);
    console.log(`📄 错误报告: ${errorPath}`);
    
    return { htmlPath: errorPath, newsCount: 0 };
  }
}

// 安装依赖并运行
async function installAndRun() {
  console.log('🔧 检查依赖...');
  
  const dependencies = ['axios', 'xml2js'];
  const missing = [];
  
  for (const dep of dependencies) {
    try {
      require.resolve(dep);
      console.log(`✅ ${dep} 已安装`);
    } catch {
      missing.push(dep);
      console.log(`❌ ${dep} 未安装`);
    }
  }
  
  if (missing.length > 0) {
    console.log(`📦 安装缺失依赖: ${missing.join(', ')}`);
    
    const { execSync } = require('child_process');
    try {
      execSync(`npm install ${missing.join(' ')}`, { 
        cwd: __dirname,
        stdio: 'inherit'
      });
      console.log('✅ 依赖安装完成');
    } catch (error) {
      console.error('❌ 依赖安装失败:', error.message);
      return;
    }
  }
  
  // 运行新闻收集
  await main();
}

// 如果直接运行此脚本
if (require.main === module) {
  installAndRun().catch(console.error);
}

module.exports = { collectRealNews, createHTMLReport };

File: real_news_collector.py
#!/usr/bin/env python3
"""
真实新闻收集器 - 使用RSS订阅收集全球科技和财经新闻
"""

import datetime
import json
import re
import time
import feedparser
import requests
import signal
from typing import List, Dict, Tuple
from urllib.parse import urlparse
import html

class RealNewsCollector:
    def __init__(self):
        self.date = datetime.datetime.now().strftime("%Y-%m-%d")
        self.tech_news = []
        self.finance_news = []
        
        # RSS源配置
        self.rss_feeds = {
            "technology": [
                ("TechCrunch", "https://techcrunch.com/feed/"),
                ("The Verge", "https://www.theverge.com/rss/index.xml"),
                ("Wired", "https://www.wired.com/feed/rss"),
                ("Ars Technica", "https://feeds.arstechnica.com/arstechnica/index"),
                ("36氪", "https://36kr.com/feed"),
                ("虎嗅", "https://www.huxiu.com/rss/0.xml"),
            ],
            "finance": [
                ("Bloomberg Markets", "https://feeds.bloomberg.com/markets/news.rss"),
                ("Reuters Business", "https://www.reutersagency.com/feed/?best-topics=business-finance&post_type=news"),
                ("CNBC", "https://www.cnbc.com/id/100003114/device/rss/rss.html"),
                ("Yahoo Finance", "https://finance.yahoo.com/news/rssindex"),
                ("界面新闻财经", "https://www.jiemian.com/lists/5/rss"),
            ]
        }
        
    def clean_html(self, text: str) -> str:
        """清理HTML标签"""
        if not text:
            return ""
        # 移除HTML标签
        text = re.sub(r'<[^>]+>', '', text)
        # 解码HTML实体
        text = html.unescape(text)
        # 清理多余空白
        text = ' '.join(text.split())
        return text.strip()
    
    def extract_summary(self, entry) -> str:
        """从RSS条目提取摘要"""
        if hasattr(entry, 'summary'):
            return self.clean_html(entry.summary)
        elif hasattr(entry, 'description'):
            return self.clean_html(entry.description)
        elif hasattr(entry, 'content'):
            if entry.content:
                content = entry.content[0].value if isinstance(entry.content, list) else str(entry.content)
                return self.clean_html(content[:200]) + "..."
        return "暂无摘要"
    
    def get_news_from_rss(self, feed_name: str, feed_url: str, category: str) -> List[Dict]:
        """从RSS源获取新闻"""
        news_items = []
        
        try:
            print(f"正在获取 {feed_name} 的新闻...")
            feed = feedparser.parse(feed_url)
            
            if feed.bozo:
                print(f"  ⚠️  解析 {feed_name} RSS源时出错：{feed.bozo_exception}")
                return news_items
            
            for entry in feed.entries[:5]:  # 每个源取最新5条
                # 提取发布时间
                pub_time = ""
                if hasattr(entry, 'published'):
                    pub_time = entry.published
                elif hasattr(entry, 'updated'):
                    pub_time = entry.updated
                
                # 格式化时间
                if pub_time:
                    try:
                        # 尝试解析时间格式
                        time_tuple = entry.get('published_parsed', entry.get('updated_parsed'))
                        if time_tuple:
                            pub_time = datetime.datetime(*time_tuple[:6]).strftime("%Y-%m-%d %H:%M")
                    except:
                        pub_time = pub_time[:50]  # 截断过长的原始时间字符串
                
                # 获取标题
                title = self.clean_html(entry.title) if hasattr(entry, 'title') else "无标题"
                
                # 跳过非今日新闻（如果是昨天的推送）
                try:
                    entry_date = datetime.datetime(*entry.published_parsed[:6]).date() if hasattr(entry, 'published_parsed') else None
                    today = datetime.datetime.now().date()
                    if entry_date and (today - entry_date).days > 1:
                        continue  # 跳过超过1天的旧闻
                except:
                    pass
                
                # 提取链接
                link = entry.link if hasattr(entry, 'link') else ""
                
                # 获取摘要
                summary = self.extract_summary(entry)
                
                # 如果摘要太长，截断
                if len(summary) > 200:
                    summary = summary[:197] + "..."
                
                news_items.append({
                    "title": title,
                    "source": feed_name,
                    "summary": summary,
                    "url": link,
                    "time": pub_time or "今天",
                    "category": category
                })
            
            print(f"  ✅ 从 {feed_name} 获取到 {len(news_items)} 条新闻")
            
        except Exception as e:
            print(f"  ❌ 获取 {feed_name} 新闻失败：{str(e)}")
        
        return news_items
    
    def collect_tech_news(self) -> List[Dict]:
        """收集科技新闻"""
        all_tech_news = []
        for feed_name, feed_url in self.rss_feeds["technology"]:
            news = self.get_news_from_rss(feed_name, feed_url, "科技")
            all_tech_news.extend(news)
            time.sleep(0.5)  # 避免请求过快
        
        # 去重（基于标题）
        seen_titles = set()
        unique_news = []
        for news in all_tech_news:
            title_lower = news["title"].lower()
            if title_lower not in seen_titles:
                seen_titles.add(title_lower)
                unique_news.append(news)
        
        # 取最新的8条
        return unique_news[:8]
    
    def collect_finance_news(self) -> List[Dict]:
        """收集财经新闻"""
        all_finance_news = []
        for feed_name, feed_url in self.rss_feeds["finance"]:
            news = self.get_news_from_rss(feed_name, feed_url, "财经")
            all_finance_news.extend(news)
            time.sleep(0.5)
        
        # 去重
        seen_titles = set()
        unique_news = []
        for news in all_finance_news:
            title_lower = news["title"].lower()
            if title_lower not in seen_titles:
                seen_titles.add(title_lower)
                unique_news.append(news)
        
        # 取最新的8条
        return unique_news[:8]
    
    def generate_html(self, tech_news: List[Dict], finance_news: List[Dict]) -> str:
        """生成HTML报告"""
        yesterday = (datetime.datetime.now() - datetime.timedelta(days=1)).strftime("%Y年%m月%d日")
        
        html = f'''<!DOCTYPE html>
<html lang="zh-CN">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>每日新闻摘要 - {self.date}</title>
    <style>
        body {{
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'PingFang SC', 'Hiragino Sans GB', 'Microsoft YaHei', sans-serif;
            line-height: 1.6;
            color: #1a1a1a;
            max-width: 1000px;
            margin: 0 auto;
            padding: 20px;
            background: linear-gradient(135deg, #f5f7fa 0%, #c3cfe2 100%);
            min-height: 100vh;
        }}
        .header {{
            text-align: center;
            padding: 25px;
            background: linear-gradient(135deg, #1e3c72 0%, #2a5298 100%);
            color: white;
            border-radius: 16px;
            margin-bottom: 30px;
            box-shadow: 0 8px 25px rgba(0, 0, 0, 0.15);
        }}
        .header h1 {{
            margin: 0;
            font-size: 32px;
            font-weight: 600;
        }}
        .date {{
            font-size: 16px;
            opacity: 0.9;
            margin-top: 10px;
            display: flex;
            align-items: center;
            justify-content: center;
            gap: 8px;
        }}
        .stats {{
            background: rgba(255, 255, 255, 0.1);
            padding: 12px 20px;
            border-radius: 12px;
            margin-top: 15px;
            display: inline-block;
        }}
        .section {{
            background: white;
            border-radius: 16px;
            padding: 30px;
            margin-bottom: 25px;
            box-shadow: 0 6px 20px rgba(0, 0, 0, 0.08);
            transition: transform 0.2s ease;
        }}
        .section:hover {{
            transform: translateY(-2px);
            box-shadow: 0 8px 25px rgba(0, 0, 0, 0.12);
        }}
        .section-title {{
            color: #1a1a1a;
            font-size: 24px;
            font-weight: 600;
            padding-bottom: 15px;
            margin-bottom: 25px;
            border-bottom: 3px solid;
            display: flex;
            align-items: center;
            gap: 12px;
        }}
        .section-title.tech {{
            border-color: #3498db;
        }}
        .section-title.finance {{
            border-color: #2ecc71;
        }}
        .news-grid {{
            display: grid;
            grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
            gap: 20px;
        }}
        .news-card {{
            background: #f8fafc;
            border-radius: 12px;
            padding: 20px;
            border: 1px solid #e2e8f0;
            transition: all 0.2s ease;
        }}
        .news-card:hover {{
            background: white;
            border-color: #cbd5e1;
            box-shadow: 0 4px 12px rgba(0, 0, 0, 0.05);
        }}
        .news-title {{
            font-size: 16px;
            font-weight: 600;
            color: #1e40af;
            margin-bottom: 10px;
            line-height: 1.4;
        }}
        .news-meta {{
            font-size: 13px;
            color: #64748b;
            margin-bottom: 12px;
            display: flex;
            align-items: center;
            gap: 10px;
            flex-wrap: wrap;
        }}
        .source-badge {{
            background: #e0f2fe;
            color: #0369a1;
            padding: 4px 10px;
            border-radius: 20px;
            font-size: 12px;
            font-weight: 500;
        }}
        .time-badge {{
            background: #f1f5f9;
            color: #475569;
            padding: 4px 10px;
            border-radius: 20px;
            font-size: 12px;
        }}
        .news-summary {{
            color: #334155;
            font-size: 14px;
            line-height: 1.5;
            margin-bottom: 15px;
        }}
        .news-link {{
            display: inline-flex;
            align-items: center;
            gap: 6px;
            color: #3b82f6;
            text-decoration: none;
            font-size: 14px;
            font-weight: 500;
            transition: color 0.2s ease;
        }}
        .news-link:hover {{
            color: #1d4ed8;
            text-decoration: underline;
        }}
        .footer {{
            text-align: center;
            margin-top: 40px;
            padding-top: 20px;
            border-top: 1px solid #e2e8f0;
            color: #64748b;
            font-size: 14px;
        }}
        .ai-badge {{
            display: inline-flex;
            align-items: center;
            gap: 6px;
            background: linear-gradient(135deg, #8b5cf6 0%, #6366f1 100%);
            color: white;
            padding: 8px 16px;
            border-radius: 20px;
            font-size: 14px;
            margin-top: 15px;
        }}
        @media (max-width: 768px) {{
            body {{
                padding: 15px;
            }}
            .header {{
                padding: 20px;
            }}
            .header h1 {{
                font-size: 24px;
            }}
            .section {{
                padding: 20px;
            }}
            .news-grid {{
                grid-template-columns: 1fr;
            }}
        }}
    </style>
</head>
<body>
    <div class="header">
        <h1>📰 每日新闻摘要</h1>
        <div class="date">
            <span>📅 {yesterday} 全球要闻</span>
            <span class="stats">🚀 {len(tech_news)} 条科技新闻 | 💰 {len(finance_news)} 条财经新闻</span>
        </div>
    </div>
    
    <div class="section">
        <h2 class="section-title tech">🚀 科技前沿</h2>
        <div class="news-grid">
        '''
        
        for news in tech_news:
            html += f'''
            <div class="news-card">
                <div class="news-title">{news['title']}</div>
                <div class="news-meta">
                    <span class="source-badge">{news['source']}</span>
                    <span class="time-badge">🕒 {news['time']}</span>
                </div>
                <div class="news-summary">{news['summary']}</div>
                <a href="{news['url']}" class="news-link" target="_blank">
                    <span>阅读原文</span>
                    <span>→</span>
                </a>
            </div>
            '''
        
        html += '''
        </div>
    </div>
    
    <div class="section">
        <h2 class="section-title finance">💰 财经动态</h2>
        <div class="news-grid">
        '''
        
        for news in finance_news:
            html += f'''
            <div class="news-card">
                <div class="news-title">{news['title']}</div>
                <div class="news-meta">
                    <span class="source-badge">{news['source']}</span>
                    <span class="time-badge">🕒 {news['time']}</span>
                </div>
                <div class="news-summary">{news['summary']}</div>
                <a href="{news['url']}" class="news-link" target="_blank">
                    <span>阅读原文</span>
                    <span>→</span>
                </a>
            </div>
            '''
        
        html += f'''
        </div>
    </div>
    
    <div class="footer">
        <div class="ai-badge">
            <span>🤖</span>
            <span>本报告由 NIKO AI助手自动生成 | 数据更新时间：{datetime.datetime.now().strftime("%Y-%m-%d %H:%M")}</span>
        </div>
        <p style="margin-top: 15px;">🔔 每天上午8:30自动推送昨日新闻摘要 | 📧 如有问题或建议请联系超级大大王</p>
    </div>
</body>
</html>
'''
        return html
    
    def save_html(self, filename: str = "daily_news_real.html"):
        """保存HTML文件"""
        print("\n" + "="*60)
        print("📡 开始收集全球新闻...")
        print("="*60)
        
        tech_news = self.collect_tech_news()
        finance_news = self.collect_finance_news()
        
        print("\n" + "="*60)
        print("📊 新闻收集完成！")
        print(f"  科技新闻：{len(tech_news)} 条")
        print(f"  财经新闻：{len(finance_news)} 条")
        print("="*60 + "\n")
        
        html_content = self.generate_html(tech_news, finance_news)
        
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(html_content)
        
        print(f"✅ HTML文件已保存：{filename}")
        return filename, tech_news, finance_news
    
    def generate_text_summary(self, tech_news: List[Dict], finance_news: List[Dict]) -> str:
        """生成文本摘要"""
        yesterday = (datetime.datetime.now() - datetime.timedelta(days=1)).strftime("%Y年%m月%d日")
        
        text = f"""📰 每日新闻摘要 - {yesterday}
────────────────────────────────

🚀 科技新闻 ({len(tech_news)}条)
"""
        
        for i, news in enumerate(tech_news, 1):
            text += f"""
{i}. {news['title']}
   📰 {news['source']} | 🕒 {news['time']}
   {news['summary']}
"""
        
        text += """
────────────────────────────────
💰 财经新闻 ({len(finance_news)}条)
"""
        
        for i, news in enumerate(finance_news, 1):
            text += f"""
{i}. {news['title']}
   📰 {news['source']} | 🕒 {news['time']}
   {news['summary']}
"""
        
        text += f"""
────────────────────────────────
📊 汇总：{len(tech_news) + len(finance_news)} 条重要新闻
💡 本摘要由 NIKO AI助手自动生成，基于多个权威新闻源
🔔 每天上午8:30自动推送昨日新闻摘要
"""
        
        return text

if __name__ == "__main__":
    collector = RealNewsCollector()
    html_file, tech_news, finance_news = collector.save_html("daily_news_real.html")
    
    # 生成文本摘要
    text_summary = collector.generate_text_summary(tech_news, finance_news)
    
    # 保存文本摘要
    with open("daily_news_summary.txt", "w", encoding="utf-8") as f:
        f.write(text_summary)
    
    print("📝 文本摘要已保存：daily_news_summary.txt")
    print("\n🎯 新闻收集任务完成！")